유저모델 생성

choitaehoon1998 opened the pull request.

choitaehoon1998 committed:
create new user model

choitaehoon1998 committed:
create new user model

choitaehoon1998 committed:
create new user model

choitaehoon1998 committed:
create new user model

choitaehoon1998 committed:
create new user model

choitaehoon1998 committed:
create new user model

choitaehoon1998 committed:
create new user model

choitaehoon1998 committed:
create new user model

choitaehoon1998 committed:
Revert "create new user model"

This reverts commit 529cdaa47326ea5d3ffab5eae2be74d30fbed107.

choitaehoon1998 committed:
Revert "create new user model"

This reverts commit 529cdaa47326ea5d3ffab5eae2be74d30fbed107.

choitaehoon1998 committed:
Revert "create new user model"

This reverts commit 529cdaa47326ea5d3ffab5eae2be74d30fbed107.

choitaehoon1998 committed:
Revert "create new user model"

This reverts commit 529cdaa47326ea5d3ffab5eae2be74d30fbed107.

choitaehoon1998 committed:
Revert "create new user model"

This reverts commit 529cdaa47326ea5d3ffab5eae2be74d30fbed107.

choitaehoon1998 committed:
Revert "create new user model"

This reverts commit 529cdaa47326ea5d3ffab5eae2be74d30fbed107.

choitaehoon1998 committed:
Revert "create new user model"

This reverts commit 529cdaa47326ea5d3ffab5eae2be74d30fbed107.

choitaehoon1998 committed:
Revert "create new user model"

This reverts commit 529cdaa47326ea5d3ffab5eae2be74d30fbed107.

choitaehoon1998 committed:
Revert "Revert "create new user model""

This reverts commit 3107d6f271a59ed918d6c7d1656e486e586fad05.

choitaehoon1998 committed:
Revert "Revert "create new user model""

This reverts commit 3107d6f271a59ed918d6c7d1656e486e586fad05.

choitaehoon1998 committed:
Revert "Revert "create new user model""

This reverts commit 3107d6f271a59ed918d6c7d1656e486e586fad05.

choitaehoon1998 committed:
Revert "Revert "create new user model""

This reverts commit 3107d6f271a59ed918d6c7d1656e486e586fad05.

choitaehoon1998 committed:
Revert "Revert "create new user model""

This reverts commit 3107d6f271a59ed918d6c7d1656e486e586fad05.

choitaehoon1998 committed:
Revert "Revert "create new user model""

This reverts commit 3107d6f271a59ed918d6c7d1656e486e586fad05.

choitaehoon1998 committed:
Revert "Revert "create new user model""

This reverts commit 3107d6f271a59ed918d6c7d1656e486e586fad05.

choitaehoon1998 committed:
Revert "Revert "create new user model""

This reverts commit 3107d6f271a59ed918d6c7d1656e486e586fad05.

choitaehoon1998 committed:
Create the application and  models

choitaehoon1998 committed:
Create the application and  models

choitaehoon1998 committed:
Create the application and  models

choitaehoon1998 committed:
Create the application and  models

choitaehoon1998 committed:
Create the application and  models

choitaehoon1998 committed:
Create the application and  models

choitaehoon1998 committed:
Create the application and  models

choitaehoon1998 committed:
Create the application and  models

choitaehoon1998 committed:
Add file "pytest.ini"

choitaehoon1998 committed:
Add file "pytest.ini"

choitaehoon1998 committed:
Add file "pytest.ini"

choitaehoon1998 committed:
Add file "pytest.ini"

choitaehoon1998 committed:
After merge

choitaehoon1998 committed:
Change settings.py os.environ -> os.environ.get

choitaehoon1998 committed:
Change settings.py os.environ -> os.environ.get

choitaehoon1998 committed:
Change settings.py to resolve mypy error

choitaehoon1998 committed:
Change settings.py to resolve mypy error

choitaehoon1998 committed:
Change settings.py to resolve mypy error

choitaehoon1998 committed:
Change settings.py to resolve mypy error

choitaehoon1998 committed:
Fix mypy error on user model

choitaehoon1998 committed:
Fix mypy error on user model

choitaehoon1998 committed:
Fix mypy error on user model

choitaehoon1998 committed:
Fix mypy error on user model

choitaehoon1998 committed:
Fix models.py import

choitaehoon1998 committed:
Fix models.py import

choitaehoon1998 committed:
Fix models.py import

choitaehoon1998 committed:
Fix models.py import

choitaehoon1998 committed:
After merge

choitaehoon1998 commented on user/models.py:
```diff
@@ -0,0 +1,79 @@
+from django.contrib.auth.models import (AbstractBaseUser, BaseUserManager,
```
지금 Run On Save 에서 나오는 텍스트들이 깨져서 import 정렬이 안되는거같아서 구글링 중입니다 이부분은 제가해결할테니 이부분만 빼고 리뷰 부탁드립니다 (-.-) => (_ _)

hyeonjames replied:
혹시나 설정에서 `source.organizeImports`를  True로 설정하지 않았나요? (이전 설정) 그랬을 확률이 높지 싶네요

elliotyim committed:
Add test db name

When pytest is executed, it creates or reuses the test db.
If the name of the test db is not specified, it will cause a key error.

elliotyim committed:
Add dependency library: drf-yasg

Drf-yasg is the library for automated API Docs which provides not only swagger
but redoc UI.

hyeonjames commented on pytest.ini:
```diff
@@ -0,0 +1,3 @@
+[pytest]
```
`pytest.ini`, `mypy.ini`, `isort.ini`, `flake8.ini`에 모든 내용이 `setup.cfg`에 있어요.
이미 추가되있는 부분이라 필요없습니다.

choitaehoon1998 replied:
넵 삭제하였습니다

hyeonjames commented on user/models.py:
```diff
@@ -0,0 +1,79 @@
+
+
+class UserManager(BaseUserManager["User"]):
+    def create_user(self, email: str, first_name: str, last_name: str,
```
`Any`는 왠만하면 안쓰는게 맞습니다. 
여긴 확실하게 `User` 타입을 쓸 수가 있으니까 그걸로 해주세요.

choitaehoon1998 replied:
이걸 제가 가장 오랫동안 고민했던 부분인데요 User 타입을 사용하기위해  -> User라고 지정 해주면 flake8에서undefined name "user"라고 나옵니다 그래서 상단에서 from user.models import User 를 해올경우에는 circular import 오류가 발생합니다 
이러한 이유로 User라는것을 알았지만 Any를 사용하였습니다

hyeonjames replied:
음?? 여기가 이미 `user.models`인데 import를 왜 해야 하나요?

choitaehoon1998 replied:
그냥 User라고 쓰면 인식을 못하던데 'User'라고 고치니까 말귀를 알아먹네요 수정했습니다

hyeonjames replied:
`User`에서 `UserManager`가 있기 때문이군요, 
그렇지만 `Any`는 좋은 대안책이 아닙니다. 오류가 나면 그냥 `'User'`로 해결할 수 있어요

hyeonjames commented on user/models.py:
```diff
@@ -0,0 +1,79 @@
+        return True
+
+
+class Photo(models.Model):
```
`Photo`는 범용적으로 쓰기로 하지 않았던가요? `user`에 들어가기 적절한 모델은 아닌거 같네요

choitaehoon1998 replied:
그럼 일단 삭제 하겠습니다!

hyeonjames commented on user/models.py:
```diff
@@ -0,0 +1,79 @@
+    USERNAME_FIELD = 'email'
+    REQUIRED_FIELDS = ['first_name', 'last_name']
+
+    def get_full_name(self) -> str:
```
이러한 함수는 `@property`을 쓰면 더 편할거에요. 
그리고 '%s%s' 는 개인 취향이겠지만, format string ( `f'{변수명}'` ) 이 더 편할거에요
```python
@property
def full_name(self) -> str:
     return f'{self.first_name}{self.last_name}'
```

choitaehoon1998 replied:
오 딱보기에도 이게더 편해보이네요 , 근데 first name, last name 컬럼을 삭제해서 이제 get_full_name 이라는 함수가 필요해 보이지않아 삭제하였습니다

hyeonjames commented on user/models.py:
```diff
@@ -0,0 +1,79 @@
+
+class UserManager(BaseUserManager["User"]):
+    def create_user(self, email: str, first_name: str, last_name: str,
+                    password: str = None, commit: bool = True) -> Any:
```
`commit`은 무엇때문에 있는 건가요?

choitaehoon1998 replied:
저도 용도를 잘느끼지 못하겠네요 삭제하겠습니다 . !

hyeonjames commented on user/models.py:
```diff
@@ -0,0 +1,79 @@
+            user.save(using=self.db)
+        return user
+
+    def create_superuser(self, email: str, first_name: str, last_name: str, password: str) -> Any:
```
마찬가지로 `Any` -> `User`

hyeonjames commented on user/models.py:
```diff
@@ -0,0 +1,79 @@
+    last_name = models.CharField(max_length=150)
+    is_active = models.BooleanField(default=True)
+    is_staff = models.BooleanField(default=False)
+    date_joined = models.DateTimeField(default=timezone.now)
```
`post`에서는 `create_date` 같은 것으로 했었죠. 마찬가지로 네이밍을 통일해주세요.
`create_date` 혹은 `joined_date` 같은 것으루요.
`auto_now_add`를 쓰는게 낫지 않을까 싶어요.

choitaehoon1998 replied:
수정하였습니다

hyeonjames commented on user/models.py:
```diff
@@ -0,0 +1,79 @@
+
+class User(AbstractBaseUser, PermissionsMixin):
+    email = models.EmailField(max_length=255, unique=True)
+    first_name = models.CharField(max_length=30)
```
영어에서는 `first_name`은 이름부분, `last_name`은 성인데요, 보통 이름 부분이 더 길지 않나요? 애매하네요.
`first_name` , `last_name`을 나눈 이유가 있나요? 없다면 그냥 `name`으로 만드는게 더 낫지 싶어요. 
@yimsb01

elliotyim replied:
저게 아마 장고에서 기본적으로 제공해 주는 User에 first_name이랑 last_name이 있어서 오버라이딩 한 것 같습니다.
물론 미국에서는 이름길이에 제한이 없다고 들었지만 제가 생각하기에도 이름 부분이 더 긴게 맞는 것 같습니다.

hyeonjames replied:
잘보니까 이 스펙은 장고에서 기본적으로 제공해주는 API 군요,
기본적으로 제공해주는 API에 대해서 굳이 오버라이딩 해주신 이유가 있나요? (DB에 생성되지 않는 다거나 그런 이유가 없다면 하지 않는게 좋겠습니다. )
`first_name`, `last_name` 같은건 API에 제공하더라도 굳이 사용하지 않아도 되는 것 같네요. 
https://docs.djangoproject.com/en/3.0/topics/auth/customizing/#specifying-custom-user-model

choitaehoon1998 replied:
first_name , last_name => name으로 변경하였습니다아

hyeonjames commented on dev_up/settings.py:
```diff
@@ -42,6 +43,7 @@
     'django.contrib.staticfiles',
     'rest_framework',
     'post',
+    'user.apps.UserConfig',
```
그냥 `user` 여야 하지 않아요?

elliotyim replied:
저게 둘 다 가능하고 저 방법이 더 좋은 이유가 이 방법이 좀 더 직접적으로 설정파일을 지정해준다고 했었던것 같은데,
전 납득이 잘 안가서 그냥 post로 적었었습니다.
그리고 통일성을 위해서 user.apps.UserConfig이 채택된다면 저도 post.apps.PostConfig로 바꾸겠습니다.

hyeonjames replied:
> 저게 둘 다 가능하고 저 방법이 더 좋은 이유가 이 방법이 좀 더 직접적으로 설정파일을 지정해준다고 했었던것 같은데,
> 전 납득이 잘 안가서 그냥 post로 적었었습니다.
> 그리고 통일성을 위해서 user.apps.UserConfig이 채택된다면 저도 post.apps.PostConfig로 바꾸겠습니다.

아하, 전 그냥 `user`, `post`가 더 이해하기 쉽고 좋을거 같아요. 직접적으로 설정파일을 지정해준 다는 뜻이 사실 확실하게 이해 가지 않네요. 동작만 한다면 가독성 측면에서 `user`, `post`로 쓰는게 더 이득이지 싶습니다.

choitaehoon1998 replied:
가독성과 통일성을 확보하기위해서 제가 "user"라고 변경하도록 하겠습니다 !

elliotyim committed:
Rollback test db name setting

choitaehoon1998 committed:
Change models.py ,settings.py

choitaehoon1998 committed:
Change models.py ,settings.py

hyeonjames commented on user/models.py:
```diff
@@ -0,0 +1,59 @@
+from django.db import models
+
+
+class UserManager(BaseUserManager["User"]):
```
이거 실행할때 오류가 날 수 밖에 없네요. 이런 경우 어쩔 수 없을 것 같네요.
`User`를 위로 옴기고, objects는 `User`가 생성될 때 초기화 하는 방식으로 가야 될거 같아요.

```python
class User( ... ):
    ...
    def __init__(self, **kwargs):
         super().__init__(**kwargs)
         self.objects = UserManager.instance  # UserManager를 Singleton 으로 만들어주세요
```

hyeonjames commented on user/models.py:
```diff
@@ -0,0 +1,59 @@
+from django.db import models
+
+
+class UserManager(BaseUserManager["User"]):
```
이거 실행할때 오류가 날 수 밖에 없네요. 이런 경우 어쩔 수 없을 것 같네요.
`User`를 위로 옴기고, objects는 `User`가 생성될 때 초기화 하는 방식으로 가야 될거 같아요.

```python
class User( ... ):
    ...
    def __init__(self, **kwargs):
         super().__init__(**kwargs)
         self.objects = UserManager.instance  # UserManager를 Singleton 으로 만들어주세요
```

choitaehoon1998 replied:
singleton 패턴 공부하고 수정해서 올릴겠습니다 좀 만 기달려주세요 !

choitaehoon1998 replied:
signleton 패턴으로 수정 한뒤 다시 review 요청 하였습니다 !

elliotyim committed:
Add --nomigrations option on setup.cfg

Without --nomigrations option, pytest will orccur an error if there is no migration file.

choitaehoon1998 committed:
Change models.py UserManager -> Singleton

choitaehoon1998 committed:
Change models.py UserManager -> Singleton

choitaehoon1998 committed:
Solve isort eroor

choitaehoon1998 committed:
Solve isort eroor

choitaehoon1998 committed:
Remove not necessary comment

choitaehoon1998 committed:
Remove not necessary comment

choitaehoon1998 committed:
Solve isort error

choitaehoon1998 committed:
Solve isort error

hyeonjames commented on user/models.py:
```diff
@@ -0,0 +1,77 @@
+    def __init__(self) -> None:
+        """ Virtually private constructor. """
+        if UserManager.__instance is not None:
+            raise Exception("This class is a singleton!")
```
별다른 `Exception`으로 줄거면 그냥 `assert` 쓰는게 나을 수 있을것 같아요.

choitaehoon1998 replied:
수정하였습니다

hyeonjames commented on user/models.py:
```diff
@@ -0,0 +1,77 @@
+    __instance = None
+
+    @staticmethod
+    def getInstance() -> Any:
```
`get_instance`로 바꿔주세요. Snake Case를 써주시길

choitaehoon1998 replied:
넵 명심하겠습니다

choitaehoon1998 committed:
Change models.py getInstance -> get_instance , exception -> assert

choitaehoon1998 committed:
Change models.py getInstance -> get_instance , exception -> assert

elliotyim committed:
Merge pull request #18 from devv-up/feature/settings

Feature/settings - Addition of test db name & dependency

elliotyim commented on user/models.py:
```diff
@@ -0,0 +1,77 @@
+from django.db import models
+
+
+class User(AbstractBaseUser, PermissionsMixin):
```
다른 이유가 있어서 그런거면 그럴수도 있지만
AbstractBaseUser는 구현해야 될게 많아서 어차피 User 모델을 상속받을거면
어느정도 구현이 되어있는 AbstractUser를 상속받는게 낫지 않나요?
is_active랑 is_staff는 사실 AbstractUser에 이미 구현이 되어 있던걸로 기억하거든요.

choitaehoon1998 replied:
저도 맨처음에 ```abstarctuser```로 할려다가 ```mypy```에서 자꾸 오류가 나서 찾아봤더니 
```abstractuser```를 사용하기위해서는 ```usermanager[abstractuser]```타입으로 정의해야 된다더군요 
근데 제가 ```usermanager[user]```타입으로 정의해줘서 ```abstractbaseuser``` 를 사용하였습니다

elliotyim replied:
아아 어쩐지..
알겠습니다 👍

hyeonjames replied:
이제 mypy에서  # type: ignore 를 적용 시켜줘서 오류가 안뜰거 같은데요, 여전히 똑같나요?

choitaehoon1998 replied:
![base](https://user-images.githubusercontent.com/51841238/76212410-bf504700-624b-11ea-8a12-e44990bfd66f.jpg)
![mypy](https://user-images.githubusercontent.com/51841238/76212417-c1b2a100-624b-11ea-99e6-05412dce8146.JPG)
abstractuser를 사용할경우 이렇게 오류가 나옵니다

choitaehoon1998 replied:
![base2](https://user-images.githubusercontent.com/51841238/76212995-0c80e880-624d-11ea-86f9-e47540ac02a0.JPG)
![mypy2](https://user-images.githubusercontent.com/51841238/76213001-10ad0600-624d-11ea-915f-c333b365e061.JPG)
objects:UserManger = UserManger() 로 수정했봤을때 결과입니다

elliotyim commented on user/admin.py:
```diff
@@ -0,0 +1,7 @@
+from user.models import User
+
+# Register your models here.
+
```
Nit: 주석 바로 밑에부터 코드를 작성하는게 좋지 않을까요?

choitaehoon1998 replied:
수정하였습니다

choitaehoon1998 replied:
어 수정한줄알았는데 admin.py 를 안올리고 models.py 만 올리고 수정했습니다 했네요

elliotyim commented on user/models.py:
```diff
@@ -0,0 +1,77 @@
+        return self.name
+
+    def has_perm(self, perm: Any, obj: Any = None) -> bool:
+
```
Nit: 위와 같은 경우입니다. 한칸 띄지 말고 붙이는게 더 나을거 같습니다.
밑에 있는 has_module_perms도요

choitaehoon1998 replied:
수정하였습니다

elliotyim committed:
Remove options: --reuse-db, --create-db

elliotyim committed:
Merge pull request #20 from devv-up/hotfix/settings

Add --nomigrations option on setup.cfg

hyeonjames commented on user/models.py:
```diff
@@ -0,0 +1,77 @@
+    def __init__(self) -> None:
+        """ Virtually private constructor. """
+        if UserManager.__instance is not None:
+            assert Exception("This class is a singleton!")
```
음? `assert`는 이렇게 쓰는게 아닙니당. 
```python
assert condition, 'error message'
```

`condition`을 통과하지 못하면, `error message` 내용으로 `AssertionError` 가 나게 하는 거에요.

```python
assert UserManager.__instance is None, "오류메세지"
```
이렇게 쓰셔야 될거 같네요.
아, 오류 메세지 같은 경우 This class 대신에 'UserManager' 라고 해주는게 좋을것 같아요. ( 뒤에 느낌표는 필요 없을 것 같아요 )

choitaehoon1998 replied:
assert를 어떻게 써야하는지 이제 정확하게 알았습니다 이제 제대로 쓸수있을거같습니다 , 
수정하였습니다

choitaehoon1998 committed:
Change models.py assert

choitaehoon1998 committed:
Change models.py assert

hyeonjames commented on user/models.py:
```diff
@@ -0,0 +1,77 @@
+        self.objects = UserManager.get_instance()
+
+
+class UserManager(BaseUserManager):  # type: ignore
```
아, `# type: ignore` 쓴 이유가 `BaseUserManager[User]` 형태가 안되서 그런건가요?
그렇다고 하면 제가 말했던 singleton 형식으로 바꿀 필요 없을 거 같아요, 죄송하지만 다시 기존 방식을 써야 될거 같네요..

choitaehoon1998 replied:
```BaseUserManager[User]```타입으로 하면 ```python manage.py makemigrations``` 명령를 실행할경우 그부분이 오류가나서 ```[User]```를 빼고 ```# type: ignore``` 를 추가해줬었습니다

choitaehoon1998 replied:
수정하였습니다

choitaehoon1998 committed:
Change models.py singleton -> singletion x

choitaehoon1998 committed:
Change models.py singleton -> singletion x

choitaehoon1998 committed:
Change incorrectly sorted

choitaehoon1998 committed:
Change incorrectly sorted

hyeonjames reviewed (approved):
👍 수고 많으셨습니다

choitaehoon1998 committed:
change admin.py , models.py line

choitaehoon1998 committed:
change admin.py , models.py line

choitaehoon1998 committed:
Change admin.py

choitaehoon1998 committed:
Change admin.py

choitaehoon1998 reviewed (commented):
![base](https://user-images.githubusercontent.com/51841238/76212410-bf504700-624b-11ea-8a12-e44990bfd66f.jpg)
![mypy](https://user-images.githubusercontent.com/51841238/76212417-c1b2a100-624b-11ea-99e6-05412dce8146.JPG)
abstractuser를 사용할경우 이렇게 오류가 나옵니다

choitaehoon1998 committed:
After merge

choitaehoon1998 committed:
Delete unnecessary file
